docs/rfc-1-staking-contract.adoc

```diff
@@ -132,47 +145,44 @@ increasing their rewards.
 
 Anyone can execute a stake top-up for an operator using a liquid T. Stake top-up
 does not automatically increase authorization levels for applications.
+Stake top-up is a one-step process and does not require any delay.
 
 ==== Undelegating a stake
 
-The owner or operator may choose to undelegate the stake by submitting an
-undeleagation order. Stake delegation should be marked in the staking contract
-as undelegating. Always the entire stake is getting undelegated and there is a
-fixed, non-governable, two months undelegation period. All applications should
-be designed in such a way that a two-month undelegation period is sufficient to
-retire all groups undelegating operators are involved in. The operator is not
-eligible for any new work during the stake undelegation period but it is still
-required to finish all the pending work. Authorized applications are eligible
-to slash the operator for the entire undelegation period.
-
-==== Recovering a stake
-
-Once the undelegation period finishes, anyone - even a third party - can recover
-the stake. Staking contract sets the stake as recovered and returns all
-delegated tokens to the owner. Applications shouldn't be able to slash after
-undelegation is complete, even before the stake is returned. Information about
-owner, operator, and beneficiary, is retained in the contract in case some
-application rewards are still available for withdrawal.
+The owner or operator may choose to undelegate the stake if there are no
+authorizations on applications or if all authorizations have been decreased to
+zero. It is not possible to partially undelegate the stake. Undelegation returns
```

Comment:
Can we allow partial undelegation? It would be useful, for example, when staker has mixed stake and want to withdraw only NU and can withdraw only partial amount. First deauthorize this amount of course.

Reply:
What I think we can do is to eliminate `undelegate(address operator) external` and replace it with:
- `unstakeKeep(address operator)` - sets cached legacy KEEP stake amount to 0; reverts if there is at least one authorization higher than the liquid T stake for that operator or if there was no KEEP staked for the operator. If succeeds, the legacy stake can be undelegated.
- `unstakeNu(address operator, uint256 amount)` - reduces cached legacy NU stake amount by `amount`; reverts if there is at least one authorization higher than the sum of remaining legacy NU stake and liquid T stake for that operator or if `amount` is higher than the cached legacy stake amount. If succeeds, the legacy NU stake can be partially undelegated.
- `unstakeT(address operator, uint256 amount)` - reduced the liquid T stake amount by `amount` and withdraws `amount` of staked T to the owner; reverts if there is at least one authorization higher than the sum of a legacy stake and remaining liquid T stake or if `amount` is higher that the liquid T stake amount.
- `unstakeAll(address operator)` - sets cached legacy stake amount to 0, sets the liquid T stake to 0 and withdraws all staked T to the owner; reverts if there is at least one non-zero authorization.

What do you think?

Reply:
Agree with idea, but few comments about naming:  in case legacy stakes `unstake` means save it from slashing but withdrawing will be outside t staking contract, when `unstake` for T includes withdraw. So maybe have different names and do not provide united method (just suggestion)?

Reply:
I'm ok with `unstake` and putting difference in docs. Also maybe some complex name like `unstakeAndWithdraw` for T?

Reply:
I was thinking about `unstakeAndWithdraw` but the complication is with `unstakeAll` that is withdrawing just T and leaving the rest of withdrawal to legacy contracts.

Reply:
`unstakeAllAndWithdrawT` ? 😄